### DISCUSSION TITLE
GUISlice example ex40_ard_ctrls_small_seesaw.ino with ard-shld-adafruit_18_joy.h config works  but not my own from GUIslice Builder does not

### DISCUSSION THREAD

**Discussion by Dowster64** [ImpulseAdventure/GUIslice#528] [Q&A] [answered]
In the working example the focus/highlight can be seen, and it can be shifted by the use of the buttons (or the joystick as originally written).  In my version I cannot see the focus shifting and I cannot select a button. I created a simple display with two GUI buttons (OK and QUIT) to test using buttons A, B and C.

From the generated code from the builder I manually added a copy of the mapping from the working example (ex40) for my buttons (in setup()) and also the max input in the header file as this is not generated by the builder:
```
#define MAX_INPUT_MAP       6
gslc_tsInputMap             m_asInputMap[MAX_INPUT_MAP];
```
I am using an Arduino Uno R4 Minima platform and Sloeber IDE. I can successfully run ex40_ard_ctrls_small_seesaw.ino.

I would really appreciate some help. I know there are some focus related functions but the example does not use them but works. So I am a little lost,
Thanks,

**Comment by Dowster64** [comment, marked as answer]
If any one else has the same issue the error was simple.  The button mapping **has** to occur after function gslc_init() has been called.  Here is the button mapping I used.  If you put it in the arduino .ino file setup() function those lines go here (after the function defined in the auto generated header from builder) :
```
// ------------------------------------------------
  // Create graphic elements
  // ------------------------------------------------
  InitGUIslice_gen();

  gslc_InitInputMap(&m_gui, m_asInputMap, MAX_INPUT_MAP); 
  gslc_InputMapAdd(&m_gui, GSLC_INPUT_PIN_ASSERT,   GSLC_PIN_BTN_A,  GSLC_ACTION_FOCUS_PREV, 0);
  gslc_InputMapAdd(&m_gui, GSLC_INPUT_PIN_ASSERT,   GSLC_PIN_BTN_B,   GSLC_ACTION_PRESELECT, 0);
  gslc_InputMapAdd(&m_gui, GSLC_INPUT_PIN_DEASSERT, GSLC_PIN_BTN_B,   GSLC_ACTION_SELECT, 0);
  gslc_InputMapAdd(&m_gui, GSLC_INPUT_PIN_ASSERT,   GSLC_PIN_BTN_C, GSLC_ACTION_FOCUS_NEXT, 0);
  gslc_InputMapAdd(&m_gui, GSLC_INPUT_PIN_ASSERT,   GSLC_PIN_BTN_UP,    GSLC_ACTION_SET_REL, -10);
  gslc_InputMapAdd(&m_gui, GSLC_INPUT_PIN_ASSERT,   GSLC_PIN_BTN_DOWN,  GSLC_ACTION_SET_REL, +10);
 // etc etc
```
I have also tried adding those lines  to the `InitGUIslice_gen()` function in the generated file after the gslc_init function call (it also works)
```
// ------------------------------------------------
// Create page elements
// ------------------------------------------------
void InitGUIslice_gen()
{
  gslc_tsElemRef* pElemRef = NULL;

  if (!gslc_Init(&m_gui,&m_drv,m_asPage,MAX_PAGE,m_asFont,MAX_FONT)) { return; }

  // Create the GUI input mapping (pin event to GUI action)
    // - The following is based on buttons that exist in the Adafruit Seesaw joystick shield
//    gslc_InitInputMap(&m_gui, m_asInputMap, MAX_INPUT_MAP);
// etc etc
```

**Comment by Pconti31** [comment]
@Dowster64 If you are doing a lot of work with this mcu combination you can make your life easier by editing the Builder's templates so thet it can inject your code automatically for any new projects.
Of course, if you are only doing one project with this mcu its no big deal.
The templates involved are all inside GUIsliceBuilder/templates
hdr.t and ino2.t
Suggested edits to hdr.t
```
// ------------------------------------------------
// Enumerations for pages, elements, fonts, images
// ------------------------------------------------
//<Enum !Start!>
//<Enum !End!>
// CODE Added for Arduino Uno R4 Minima
#define MAX_INPUT_MAP       6
gslc_tsInputMap             m_asInputMap[MAX_INPUT_MAP];
```
Suggested edits to ino2.t
```
  // ------------------------------------------------
  // Create graphic elements
  // ------------------------------------------------
  InitGUIslice_gen();
// CODE Added for Arduino Uno R4 Minima
  gslc_InitInputMap(&m_gui, m_asInputMap, MAX_INPUT_MAP); 
  gslc_InputMapAdd(&m_gui, GSLC_INPUT_PIN_ASSERT,   GSLC_PIN_BTN_A,  GSLC_ACTION_FOCUS_PREV, 0);
  gslc_InputMapAdd(&m_gui, GSLC_INPUT_PIN_ASSERT,   GSLC_PIN_BTN_B,   GSLC_ACTION_PRESELECT, 0);
  gslc_InputMapAdd(&m_gui, GSLC_INPUT_PIN_DEASSERT, GSLC_PIN_BTN_B,   GSLC_ACTION_SELECT, 0);
  gslc_InputMapAdd(&m_gui, GSLC_INPUT_PIN_ASSERT,   GSLC_PIN_BTN_C, GSLC_ACTION_FOCUS_NEXT, 0);
  gslc_InputMapAdd(&m_gui, GSLC_INPUT_PIN_ASSERT,   GSLC_PIN_BTN_UP,    GSLC_ACTION_SET_REL, -10);
  gslc_InputMapAdd(&m_gui, GSLC_INPUT_PIN_ASSERT,   GSLC_PIN_BTN_DOWN,  GSLC_ACTION_SET_REL, +10);
```
Just an FYI
Paul--

**Reply by Dowster64** [reply]
Thanks @Pconti31, that was a really useful insight.  It should help people with the Adafruit 1.8" Shield,
Thanks again.
